### PR TITLE
spark: add GCP run and job facets

### PR DESCRIPTION
**Pull request by codelixir**
### Problem

Adds GCPRunFacetBuilder and GCPJobFacetBuilder to collect GCP-specific properties at run and job levels respectively.

Closes: #2641

### Solution

The `SPARK_DIST_CLASSPATH` property can be used to determine whether it is a GCP Dataproc environment, depending on which we can decide to add the respective facet builders to the list. Other properties are obtained from Google's compute metadata API.

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [x] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

#### One-line summary:

Adds GCPRunFacetBuilder and GCPJobFacetBuilder to report additional facets when running on Google Cloud Platform

### Checklist

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [x] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [ ] Your changes are accompanied by tests (_if relevant_)
- [ ] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] Your comment includes a one-liner for the changelog about the specific purpose of the change (_if necessary_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)
- [x] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2023 contributors to the OpenLineage project

### CODE REVIEW

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/EnvironmentFacet.java [hunk 1]
```diff
@@ -15,7 +15,7 @@
  * cluster used, reporting certain environment variables, or resolving mount points.
  */
 public class EnvironmentFacet extends OpenLineage.DefaultRunFacet {
-  @JsonProperty("environment-properties")
+  @JsonIgnore
```
Comment (pawel-big-lebowski) [inline integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/EnvironmentFacet.java:18]: why is this necessary?

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            .build();
+    HTTP_CLIENT = HttpClients.custom().setDefaultRequestConfig(config).build();
+  }
+
+  public static boolean isGCPRuntime() {
+    String sparkDistClasspath = Environment.getEnvironmentVariable(SPARK_DIST_CLASSPATH);
+    return (sparkDistClasspath != null && sparkDistClasspath.contains(DATAPROC_CLASSPATH));
+  }
+
+  public static void setDataprocSpecificFacets(
```
Comment (pawel-big-lebowski) [inline integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java:84]: wouldn't be better for this method to return a map instead filling the map provided as argument? If not, it may be worth adding feature flag in config enabling this to be turned ON/OFF.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.apache.spark.SparkContext;
+import org.apache.spark.sql.execution.SparkPlan;
+import org.apache.spark.sql.execution.WholeStageCodegenExec;
+
+/** Util to extract values from GCP environment */
+public class GCPUtils {
+
+  private static final String PROJECT_ID_URI =
```
Comment (pawel-big-lebowski) [inline integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java:35]: Isn't there an endpoint to return all the required information. This would allow avoiding three separate http calls to get what is needed.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        getDriverHost(sparkContext).ifPresent(p -> dataprocProperties.put("spark.driver.host", p));
+        getClusterName(sparkContext)
+            .ifPresent(p -> dataprocProperties.put("spark.cluster.name", p));
+        getRegionName().ifPresent(p -> dataprocProperties.put("spark.cluster.region", p));
+        getClusterUUID().ifPresent(p -> dataprocProperties.put("spark.cluster.uuid", p));
+        getGCPJobID(sparkContext).ifPresent(p -> dataprocProperties.put("spark.job.id", p));
+        getGCPJobUUID(sparkContext).ifPresent(p -> dataprocProperties.put("spark.job.uuid", p));
+        break;
+      case BATCH:
+        getRegionName().ifPresent(p -> dataprocProperties.put("spark.batch.region", p));
```
Comment (pawel-big-lebowski) [inline integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java:99]: I assume endpoints return JSONs and here we put them into string fields. Do you really need whole json content? Wouldn't it make sense to extract exactly the fields you need.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.apache.spark.SparkContext;
+import org.apache.spark.sql.execution.SparkPlan;
+import org.apache.spark.sql.execution.WholeStageCodegenExec;
+
+/** Util to extract values from GCP environment */
+public class GCPUtils {
```
Comment (pawel-big-lebowski) [inline integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java:33]: I would appreciate writing some tests for this class. You can start MockServer that responds with the jsons and verify if the class methods work as expected.

**Review by pawel-big-lebowski — COMMENTED** [review]
Great piece of code. Thank you for contributing. Please look at my questions and comments below.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java (hunk 3 above)
Comment (codelixir) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java:35]: Unfortunately, there isn't a single endpoint that returns everything. For example, `http://metadata.google.internal/computeMetadata/v1/instance/attributes` returns just a list of the keys available for that instance (`dataproc-region`, `dataproc-cluster-uuid` etc.)
Although these values are being cached, as far as I am aware.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java (hunk 2 above)
Comment (codelixir) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java:84]: Right, will change this to return a map instead.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java (hunk 4 above)
Comment (codelixir) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java:99]: These endpoints return a single value (corresponding to the field) and not a JSON.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/EnvironmentFacet.java (hunk 1 above)
Comment (codelixir) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/EnvironmentFacet.java:18]: For now, the environment-properties facet looks something like this 
```
"environment-properties": {
  "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.13.0-SNAPSHOT/integration/spark",
  "_schemaURL": "https://openlineage.io/spec/2-0-2/OpenLineage.json#/$defs/RunFacet",
  "environment-properties": {
    "foo": "...",
    "bar": "..."
  }
}
```
This change will just make a cleaner JSON:
```
"environment-properties": {
  "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.13.0-SNAPSHOT/integration/spark",
  "_schemaURL": "https://openlineage.io/spec/2-0-2/OpenLineage.json#/$defs/RunFacet",
  "foo": "...",
  "bar": "..."
}
```
Please let me know if it would be okay to keep the change. If not, I will revert it.

Comment (codelixir) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/EnvironmentFacet.java]: I have removed this and moved to a different approach for reporting facets.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java (hunk 2 above)
Comment (codelixir) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java]: Done

**Comment by codelixir** [comment]
Hi, I have modified the run facet according to #2771 and also added the job facet proposed in #2740.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java (hunk 5 above)
Comment (codelixir) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java:33]: Okay, working on this.

Comment (codelixir) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/GCPUtils.java:33]: Hi @pawel-big-lebowski, I have added tests for this in the recent commit. Please have a look whenever convenient.

**Comment by codelixir** [comment]
The CI tests which are failing were taken care of in #2655

**Review by pawel-big-lebowski — APPROVED** [review]
Great piece of work. Thank you for contributing.

**Comment by pawel-big-lebowski** [comment]
> The CI tests which are failing were taken care of in #2655

could you then rebase your changes?
